Add Debugger+OpernFermion Sections (#5941)
PennyLaneAI/pennylane: doc/releases/changelog-0.37.0.md

  analyzing quantum circuits during execution.
  [(#5680)](https://github.com/PennyLaneAI/pennylane/pull/5680)
  [(#5749)](https://github.com/PennyLaneAI/pennylane/pull/5749)
  [(#5789)](https://github.com/PennyLaneAI/pennylane/pull/5789)
  
  This includes monitoring the circuit via measurements using `qml.debug_state()`, `qml.debug_probs()`, 
  `qml.debug_expval()`, and `qml.debug_tape()`, stepping through the operations in a quantum circuit, and interactively
  adding operations during execution.

  Including `qml.breakpoint()` in a circuit will cause the simulation to pause during execution and bring up the interactive console.
  For example, consider the following code in a Python file called `script.py`:

  ```python
  @qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
  def circuit(x):
      qml.Hadamard(wires=0)
      qml.CNOT(wires=(0,2))
      qml.breakpoint()

      qml.RX(x, wires=1)
      qml.RY(x, wires=2)
      qml.breakpoint()

      return qml.sample()

  circuit(1.2345)
  ```

  Upon executing `script.py`, the simulation pauses at the first breakpoint:

  ```pycon
  > /Users/your/path/to/script.py(8)circuit()

>   -> qml.RX(-x, wires=1)

Reviewer: ```suggestion
  -> qml.RX(x, wires=1)
```

  (pldb):
  ```

  While debugging, we can access circuit information.
  `qml.debug_tape()` returns the tape of the circuit, giving access to its operations and drawing:

  ```pycon

>   (pldb): tape = qml.debug_tape()

Reviewer: @Jaybsoni why did we go with `[pldb]` using square brackets rather than rounded brackets?

  (pldb): print(tape.draw(wire_order=[0,1,2]))
  0: ──H─╭●─┤  
  2: ────╰X─┤  
  (pldb): tape.operations
  [Hadamard(wires=[0]), CNOT(wires=[0, 2])]
  ```

  `qml.debug_state()` is equivalent to `qml.state()` and gives the current state:

  ```pycon
  (pldb): print(qml.debug_state())
  [0.70710678+0.j 0.        +0.j 0.        +0.j 0.        +0.j
    1.        +0.j 0.70710678+0.j 0.        +0.j 0.        +0.j]
  ```

  Other debugger functions like `qml.debug_probs()` and `qml.debug_expval()` also function like their simulation counterparts (`qml.probs` and `qml.expval`, respectively) and are described in more detail in the [debugger documentation](https://docs.pennylane.ai/en/stable/code/qml_debugging.html)
  

>   Additionally, standard debugging commands are available to navigate through code, including `list`, `longlist`, `next`, `continue`, and `quit` as described in [the debugging documentation](https://docs.pennylane.ai/en/stable/code/qml_debugging.html#controlling-code-execution-in-the-debugging-context).

Reviewer: ```suggestion
  Additionally, [standard debugging commands are available](https://docs.pennylane.ai/en/stable/code/qml_debugging.html#controlling-code-execution-in-the-debugging-context)
  to navigate through code, including `list`, `longlist`, `next`, `continue`, and `quit`.
```
Not a blocker, it just felt too similar to line 230 in how the docs were referenced at the end.

  
  Finally, to modify a circuit mid-run, simply call the desired PennyLane operations:

  ```pycon
  (pldb) qml.CNOT(wires=(0,2))
  CNOT(wires=[0, 2])
  (pldb): print(qml.debug_tape().draw(wire_order=[0,1,2]))
  0: ──H─╭●─╭●─┤  
  2: ────╰X─╰X─┤  
  ```

<h4>Convert between OpenFermion and PennyLane 🤝</h4>

* Two new functions called `qml.from_openfermion` and `qml.to_openfermion` are now available to convert between 
  OpenFermion and PennyLane objects. This includes both fermionic and qubit operators.
  [(#5773)](https://github.com/PennyLaneAI/pennylane/pull/5773)
  [(#5808)](https://github.com/PennyLaneAI/pennylane/pull/5808)
  [(#5881)](https://github.com/PennyLaneAI/pennylane/pull/5881)
  
  For fermionic operators:

  ```pycon
  >>> of_fermionic = openfermion.FermionOperator('0^ 2')

>   >>> type(of_fermionic)
>   <class 'openfermion.ops.operators.fermion_operator.FermionOperator'>

Reviewer: ```suggestion
```

  >>> pl_fermionic = qml.from_openfermion(of_fermionic)

>   >>> type(pl_fermionic)
>   <class 'pennylane.fermi.fermionic.FermiWord'>

Reviewer: ```suggestion
```

  To learn more about this new functionality, check out our [noise module documentation](https://docs.pennylane.ai/en/stable/code/qml_noise.html)
  and keep your eyes peeled for an in-depth demo!

<h4>Catch bugs with the PennyLane debugger 🚫🐞</h4>

* The new PennyLane quantum debugger (`PLDB`) allows pausing simulation via the `qml.breakpoint()` command and provides tools for 
  analyzing quantum circuits during execution.
  [(#5680)](https://github.com/PennyLaneAI/pennylane/pull/5680)
  [(#5749)](https://github.com/PennyLaneAI/pennylane/pull/5749)
  [(#5789)](https://github.com/PennyLaneAI/pennylane/pull/5789)
  
  This includes monitoring the circuit via measurements using `qml.debug_state()`, `qml.debug_probs()`, 
  `qml.debug_expval()`, and `qml.debug_tape()`, stepping through the operations in a quantum circuit, and interactively
  adding operations during execution.

>   Including `qml.breakpoint()` in a circuit will cause the simulation to pause during execution and bring up the `[pldb]` interactive console:

Reviewer 2: ```suggestion
  Including `qml.breakpoint()` in a circuit will cause the simulation to pause during execution and bring up the `[pldb]` interactive console.
  For example, consider the following code in a Python file called `script.py`:
```

  ```python
  @qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
  def circuit(x):
      qml.Hadamard(wires=0)
      qml.CNOT(wires=(0,2))
      qml.breakpoint()

      qml.RX(-x, wires=1)
      qml.RY(-x, wires=2)
      qml.breakpoint()

      return qml.sample()

  circuit(1.2345)
  ```

>   Upon execution, the simulation pauses at the first breakpoint:

Reviewer 2: ```suggestion
  Upon executing `script.py`, the simulation pauses at the first breakpoint:
```

  ```pycon
  > /Users/your/path/to/script.py(8)circuit()
  -> qml.RX(-x, wires=1)
  [pldb]:
  ```
  While debugging, we can access circuit information:
  - `qml.debug_tape()` - returns the tape of the circuit, giving access to its operations and drawing.
    ```pycon
    [pldb]: tape = qml.debug_tape()
    [pldb]: print(tape.draw(wire_order=[0,1,2]))
    0: ──H─╭●─────┤  
    1: ────│───RX─┤  
    2: ────╰X──RY─┤ 
    [pldb]: tape.operations
    [Hadamard(wires=[0]), CNOT(wires=[0, 2])]
    ```
  - `qml.debug_state()` - equivalent to `qml.state()`, gives the current state.
    ```pycon
    [pldb]: print(qml.debug_state())
    [0.70710678+0.j 0.        +0.j 0.        +0.j 0.        +0.j
     1.        +0.j 0.70710678+0.j 0.        +0.j 0.        +0.j]
    ```

>   Other debugger functions like `qml.debug_probs()` and `qml.debug_expval()` also function like their simulation counterparts and are described in more detail in the [debugger documentation](https://docs.pennylane.ai/en/stable/code/qml_debugging.html)

Reviewer 2: ```suggestion
  Other debugger functions like `qml.debug_probs()` and `qml.debug_expval()` also function like their simulation counterparts (`qml.probs` and `qml.expval`, respectively) and are described in more detail in the [debugger documentation](https://docs.pennylane.ai/en/stable/code/qml_debugging.html)
```

  - `list` or `longlist` - view the current position and surrounding code.
    ```pycon
    [pldb]: longlist
     2     @qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
     3     def circuit(x):
     4         qml.Hadamard(wires=0)
     5         qml.CNOT(wires=(0,2))
     6         qml.breakpoint()
     7  
     8  ->     qml.RX(-x, wires=1)
     9         qml.RY(-x, wires=2)
    10         qml.breakpoint()
    11  
    12         return qml.sample()
    ```
  - `next` - apply the next operation in the circuit.
    ```pycon
    [pldb]: next
    > /Users/your/path/to/script.py(9)circuit()
    -> qml.RY(-x, wires=2)
    ```
  - `continue` - continue to execute the code until another breakpoint is reached.
    ```pycon
    [pldb]: continue
    > /Users/your/path/to/script.py(12)circuit()
    -> return qml.sample()
    ```
  - `quit` - exit debugging.
  
  Finally, to modify a circuit mid-run, simply call the desired PennyLane operations:

  ```pycon
  > /Users/your/path/to/script.py(12)circuit()
  -> return qml.sample()
  [pldb] qml.CNOT(wires=(0,2))
  CNOT(wires=[0, 2])
  [pldb]: print(qml.debug_tape().draw(wire_order=[0,1,2]))
  0: ──H─╭●─────╭●─┤  
  1: ────│───RX─│──┤  

>   2: ────╰X──RY─╰X─┤ 

Reviewer 2: Nice! 🎉

  ```

<h4>Convert between OpenFermion and PennyLane 🤝</h4>

> * The ``from_openfermion`` and ``to_openfermion`` are now available to convert between 

Reviewer 2: ```suggestion
* Two new functions called `qml.from_openfermion` and `qml.to_openfermion` are now available to convert between 
```

  OpenFermion and PennyLane objects. This includes both fermionic and qubit operators.
  [(#5773)](https://github.com/PennyLaneAI/pennylane/pull/5773)
  [(#5808)](https://github.com/PennyLaneAI/pennylane/pull/5808)
  [(#5881)](https://github.com/PennyLaneAI/pennylane/pull/5881)

  For qubit operators:

>   ```pycon
>   >>> of_qubit = 0.5*openfermion.QubitOperator('X0 X5')

Reviewer 2: ```suggestion
  ```pycon
  >>> import openfermion
  >>> of_qubit = 0.5 * openfermion.QubitOperator('X0 X5')
```

  >>> type(of_qubit)
  <class 'openfermion.ops.operators.qubit_operator.QubitOperator'>
  >>> pl_qubit = qml.from_openfermion(of_qubit)
  >>> type(pl_qubit)
  <class 'pennylane.ops.op_math.linear_combination.LinearCombination'>
  >>> print(pl_qubit)
  0.5 * (X(0) @ X(5))
  ```
  

>   And for Fermionic operators:

Reviewer 2: ```suggestion
  And for fermionic operators:
```

  To learn more about this new functionality, check out our [noise module documentation](https://docs.pennylane.ai/en/stable/code/qml_noise.html)
  and keep your eyes peeled for an in-depth demo!

<h4>Catch bugs with the PennyLane debugger 🚫🐞</h4>

> * The new PennyLane quantum debugger (`PLDB`) allows pausing simulation via the `qml.breakpoint()` command and provides tools for 

Reviewer: How much do we want to market PLDB as a thing? I'm not sure we need to so much :thinking:

Reviewer 2: This is a good question. It's kinda making it look like its own standalone product if we give it an acronym 😅. But, I see the use for it because the UI has `pldb` all over the place. @DSGuala what about if we just make it lower-case? 🤔 

```suggestion
* The new PennyLane quantum debugger (`pldb`) allows pausing simulation via the `qml.breakpoint()` command and provides tools for 
```

Author: Decided to just remove it.

Reviewer: I still see it :thinking: not a blocker though.

  `qml.debug_expval()`, and `qml.debug_tape()`, stepping through the operations in a quantum circuit, and interactively
  adding operations during execution.

  Including `qml.breakpoint()` in a circuit will cause the simulation to pause during execution and bring up the `[pldb]` interactive console.
  For example, consider the following code in a Python file called `script.py`:

  ```python
  @qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
  def circuit(x):
      qml.Hadamard(wires=0)
      qml.CNOT(wires=(0,2))
      qml.breakpoint()

      qml.RX(-x, wires=1)
      qml.RY(-x, wires=2)
      qml.breakpoint()

      return qml.sample()

  circuit(1.2345)
  ```
  Upon executing `script.py`, the simulation pauses at the first breakpoint:
  ```pycon
  > /Users/your/path/to/script.py(8)circuit()
  -> qml.RX(-x, wires=1)
  [pldb]:
  ```
  While debugging, we can access circuit information:
  - `qml.debug_tape()` - returns the tape of the circuit, giving access to its operations and drawing.
    ```pycon
    [pldb]: tape = qml.debug_tape()
    [pldb]: print(tape.draw(wire_order=[0,1,2]))
    0: ──H─╭●─────┤  
    1: ────│───RX─┤  

>     2: ────╰X──RY─┤ 

Reviewer: I got this:
```
0: ──H─╭●─┤  
2: ────╰X─┤  
```

  
  Additionally, standard debugging commands are available to navigate through code:
  - `list` or `longlist` - view the current position and surrounding code.
    ```pycon
    [pldb]: longlist
     2     @qml.qnode(qml.device('default.qubit', wires=(0,1,2)))
     3     def circuit(x):
     4         qml.Hadamard(wires=0)
     5         qml.CNOT(wires=(0,2))
     6         qml.breakpoint()
     7  
     8  ->     qml.RX(-x, wires=1)
     9         qml.RY(-x, wires=2)
    10         qml.breakpoint()
    11  
    12         return qml.sample()
    ```
  - `next` - apply the next operation in the circuit.
    ```pycon
    [pldb]: next
    > /Users/your/path/to/script.py(9)circuit()
    -> qml.RY(-x, wires=2)
    ```
  - `continue` - continue to execute the code until another breakpoint is reached.
    ```pycon
    [pldb]: continue
    > /Users/your/path/to/script.py(12)circuit()
    -> return qml.sample()
    ```

>   - `quit` - exit debugging.

Reviewer: This is an area we could shorten, e.g., by linking to existing Python docs?

  
  Finally, to modify a circuit mid-run, simply call the desired PennyLane operations:

  ```pycon
  > /Users/your/path/to/script.py(12)circuit()
  -> return qml.sample()
  [pldb] qml.CNOT(wires=(0,2))
  CNOT(wires=[0, 2])
  [pldb]: print(qml.debug_tape().draw(wire_order=[0,1,2]))
  0: ──H─╭●─────╭●─┤  
  1: ────│───RX─│──┤  
  2: ────╰X──RY─╰X─┤ 
  ```

<h4>Convert between OpenFermion and PennyLane 🤝</h4>

* Two new functions called `qml.from_openfermion` and `qml.to_openfermion` are now available to convert between 
  OpenFermion and PennyLane objects. This includes both fermionic and qubit operators.
  [(#5773)](https://github.com/PennyLaneAI/pennylane/pull/5773)
  [(#5808)](https://github.com/PennyLaneAI/pennylane/pull/5808)
  [(#5881)](https://github.com/PennyLaneAI/pennylane/pull/5881)

  For qubit operators:
  ```pycon
  >>> import openfermion
  >>> of_qubit = 0.5 * openfermion.QubitOperator('X0 X5')

>   >>> type(of_qubit)

Reviewer: Not a blocker, but the type information isn't _that_ useful from my perspective.

  <class 'openfermion.ops.operators.qubit_operator.QubitOperator'>
  >>> pl_qubit = qml.from_openfermion(of_qubit)
  >>> type(pl_qubit)
  <class 'pennylane.ops.op_math.linear_combination.LinearCombination'>
  >>> print(pl_qubit)
  0.5 * (X(0) @ X(5))
  ```
  

>   And for fermionic operators:

Reviewer: What about mentioning fermionic ops first? Since I think most users will have that as the starting point
